packages/logging/src/analytics-helpers.ts

```diff
@@ -119,4 +159,206 @@ export class ToggleableAnalytics implements MongoshAnalytics {
       }
     }
   }
+
+  flush(callback: (err?: Error | undefined) => void): void {
+    return this._target.flush(callback);
+  }
+}
+
+type ThrottledAnalyticsOptions = {
+  target: MongoshAnalytics;
+  /**
+   * Throttling options. If not provided, throttling is disabled (default: null)
+   */
+  throttle: {
+    /** Allowed events per timeframe number */
+    rate: number;
+    /** Timeframe for throttling in milliseconds (default: 60_000ms) */
+    timeframe?: number;
+    /** Path to persist rpm value to be able to track them between sessions */
+    metadataPath: string;
+    /** Duration in milliseconds in which the lock is considered stale (default: 43_200_000) */
+    lockfileStaleDuration?: number;
+  } | null;
+};
+
+async function lockfile(
+  filepath: string,
+  staleDuration = 43_200_000
+): Promise<() => Promise<void>> {
+  let intervalId: ReturnType<typeof setInterval>;
+  const lockfilePath = `${filepath}.lock`;
+  const unlock = async() => {
+    clearInterval(intervalId);
+    try {
+      return await fs.promises.rmdir(lockfilePath);
+    } catch {
+      // ignore update errors
+    }
+  };
+  try {
+    await fs.promises.mkdir(lockfilePath);
```

Comment:
is there a particular reason for using a directory over an empty file?

Reply:
This seems to be the most common way to create lockfiles as mkdir is an atomic operation even on windows, compared to any ways of opening or writing to a file that are hard to lock to a single process. Am I missing something simpler? 🤔